Various internal MultiIndex improvements

vyasr opened the pull request:
<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please mark your pull request as Draft.
   https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove it from "Draft" and make it "Ready for Review".
   https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#marking-a-pull-request-as-ready-for-review

   If assistance is required to complete the functionality, for example when the
   C/C++ code of a feature is complete but Python bindings are still required,
   then add the label `help wanted` so that others can triage and assist.
   The additional changes then can be implemented on top of the same PR.
   If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on the target branch, force push, or rewrite history.
   Doing any of these causes the context of any comments made by reviewers to be lost.
   If conflicts occur against the target branch they should be resolved by
   merging the target branch into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->
This PR is a follow-up to #9191 to further simplify the MultiIndex class. It removes various unused functions, inlines and simplifies a number of other single-use internal functions, and moves more methods to Frame that can be shared. It also makes numerous other miscellaneous improvements to the code in MultiIndex to simplify further rewrites.

vyasr commented:
Two notes:
- I've marked this as breaking because the `labels` parameter was removed from the constructor. This parameter has been deprecated for over two years and should be safe to remove at this point.
- I've targeted this at 21.10 to take advantage of CI for the moment but we can safely retarget to 21.12 if we don't want to squeeze this in at this point. There's nothing critical in this PR.

galipremsagar commented on python/cudf/cudf/core/frame.py:
```diff
@@ -4718,17 +4729,8 @@ def shape(self):
         return (len(self),)
 
     def __iter__(self):
-        """
```
Do we want to remove this docstring?

vyasr replied:
I think it's fine either way. I can add it back if you prefer. I think this behavior would be cleaner to move to a mixin anyway, in which case it would be much more self-documenting in the long term and easier to add to different classes.

vyasr replied:
Discussed offline, I added it back to be safe.

thomcom commented on python/cudf/cudf/core/multiindex.py:
```diff
@@ -269,10 +262,6 @@ def _from_data(
             obj.name = name
         return obj
```
Way to navigate the madness of levels/codes prep.

thomcom replied:
:+1:

thomcom commented on python/cudf/cudf/core/multiindex.py:
```diff
@@ -1539,13 +1418,9 @@ def memory_usage(self, deep=False):
-            temp_other = self.to_pandas()
-        return temp_self.difference(temp_other, sort)
+            other = other.to_pandas()
+        return self.to_pandas().difference(other, sort)
```
I see this function is never tested, or used. Any knowledge of why it is here?

vyasr replied:
`Index.difference` exists in pandas, so this method probably should exist. The fact that tests are missing is probably why this clearly buggy implementation wasn't caught before, I can add some tests in this PR.

thomcom reviewed (commented):
A delightful stroll down memory lane. Great job digging through the miserable special cases of `MultiIndex`, thanks!

vyasr commented:
Temporarily leaving this PR pointing to branch-21.10 to for CI, but moving to the 21.12 milestone. Will retarget the PR once changes are finalized (or 21.12 CI is up and running).

marlenezw commented on python/cudf/cudf/core/_base_index.py:
```diff
@@ -162,6 +167,38 @@ def _clean_nulls_from_index(self):
+        return self.is_monotonic_increasing
+
+    @property
+    def is_monotonic_increasing(self):
```
Why can we use `self.is_monotonic_increasing` above but then here we're raising a not implemented error?

vyasr replied:
We want `is_monotonic` to be an alias for `is_monotonic_increasing` for all types of indexes. The problem is that the implementation of `is_monotonic_increasing` is different for different types. By using `is_monotonic_increasing` above but making it raise here, we ensure that for any index that defines `is_monotonic_increasing` we will observe the correct behavior for `is_monotonic` without duplicating the logic, but we will get an error for subclasses that don't override this method.

marlenezw commented:
Overall this is looking good to me. It might be able to make it into 21.10 if there aren't any other major changes (not sure when burn down is though.)

galipremsagar commented on python/cudf/cudf/core/multiindex.py:
```diff
@@ -78,72 +77,72 @@ def __init__(
-            if -1 in self._codes[n].values:
+        source_data = {}
+        for i, (cname, col) in enumerate(codes._data.items()):
+            if -1 in col.values:
```
Instead of this, should calling `cudf::contains` be faster? 
https://github.com/rapidsai/cudf/blob/branch-21.10/cpp/include/cudf/search.hpp#L138

We might have to expose the api to cython bindings

galipremsagar commented on python/cudf/cudf/core/multiindex.py:
```diff
@@ -708,11 +636,6 @@ def isin(self, values, level=None):
 
         return result
 
-    def mask(self, cond, other=None, inplace=False):
```
Removing this will let users hit `Frame.mask`, is it an intentional change?

vyasr commented on python/cudf/cudf/core/multiindex.py:
```diff
@@ -78,72 +77,72 @@ def __init__(
-            codes[0], (Sequence, np.ndarray)
-        ):
-            raise TypeError("Codes is not a Sequence of sequences")
+        levels = [cudf.Series(level) for level in levels]
```
I agree. I plan to make more extensive changes to the fundamental MultiIndex data structures in a subsequent PR so I didn't touch the definition of levels here, but I'll definitely be aiming to change that in the future.

vyasr commented on python/cudf/cudf/core/multiindex.py:
```diff
@@ -78,72 +77,72 @@ def __init__(
-            codes = as_index(self._codes[n]._column)
-            if -1 in self._codes[n].values:
+        source_data = {}
+        for i, (cname, col) in enumerate(codes._data.items()):
```
Done.

vyasr commented on python/cudf/cudf/core/multiindex.py:
```diff
@@ -78,72 +77,72 @@ def __init__(
-            source_data[n] = libcudf.copying.gather(
-                level, codes._data.columns[0]
-            )[0][n]
+            source_data[cname] = libcudf.copying.gather(level, col)[0][cname]
```
Yeah, I think right now I'm going to focus on simplifications and then come back to questions of performance. I don't know enough yet to understand under what circumstances a MultiIndex is useful just in terms of its codes and levels. If there are many operations that don't need the full `_data` dict to be materialized we can change that, but I should note that that is actually a different optimization than what was being done before and would be significantly more useful I think.

vyasr commented:
@galipremsagar once you're happy with this it's up to you whether you want to merge this into 21.10 or 21.12. It seems like it's basically ready, but as we discussed offline (and commented on this thread above) it may be wiser to hold off until 21.12 to avoid any last minute surprises.

galipremsagar reviewed (approved):
LGTM, lets merge this into `branch-21.12` as discussed offline. Can you rebase the changes?

vyasr commented:
Wow I completely borked the merge somehow. Going to rebase, will need to remove unnecessary requested reviewers.

vyasr commented:
Ah I think I see the problem. Has forward-merge from 21.10->21.12 not been set up yet? I think I had commits from merging 21.10 in my PR that were not yet present on 21.12.

vyasr commented:
I had fixed a typo in a test created in #9258, which led to conflicts from merges. I've removed that change entirely to avoid causing issues with the 21.10->21.12 forward merge once it's introduced.

vyasr commented:
@gpucibot merge
